File: Open3D-master/examples/python/solution.py
import os
import open3d as o3d

def write_to_file(list_of_lists, file_path):
    output = ""
    for lst in list_of_lists:
        for ele in lst:
            output += f"{ele}\t"
        output += "\n"
    with open(file_path, "w") as fp:
        fp.write(output)

def main():
    MESH_PATH = "../test_mesh.ply"
    OUTPUT_FILE_PATH = "examples/result.txt"

    assert os.path.exists(MESH_PATH)
    mesh = o3d.io.read_triangle_mesh(MESH_PATH)

    if not mesh.has_adjacency_list(): 
        mesh.compute_adjacency_list()
        # Q1. Do I need to compute adjacency matrix inside core C++ `IdenticallyColoredConnectedComponents`
        # instead of `mesh.compute_adjacency_list()` as above.

    connected_components = mesh.identically_colored_connected_components()
    # Q2. Do I need to sort lists and their elements inside `connected_components`?
    #   - sort by size of lists inside `connected_components`
    #   - sort inside elements of lists inside `connected_components`
    #
    # If YES, do I need to make changes in core C++ code / example code?
    
    write_to_file(connected_components, OUTPUT_FILE_PATH)


if __name__ == "__main__":
    main()

File: Open3D-master/examples/cpp/Solution.cpp
#include <iostream>
#include <fstream>
#include "open3d/Open3D.h"

int main(int argc, char *argv[]) {
    using namespace open3d;

    std::string MESH_PATH = "../test_mesh.ply";
    std::string OUTPUT_FILE = "./examples/result.txt";

    geometry::TriangleMesh mesh;
    io::ReadTriangleMeshOptions options;
    bool success = io::ReadTriangleMeshFromPLY(MESH_PATH, mesh, options);
    if (!success){
        std::cout<<"[!!!] please make sure test_mesh.ply exists at the given `MESH_PATH`.";
    }

    if (!mesh.HasAdjacencyList()){
        mesh.ComputeAdjacencyList();
    }    
    std::vector<std::vector<long unsigned int>> all_ccs = mesh.IdenticallyColoredConnectedComponents();

    std::string output; 
    for (long unsigned int i=0; i<all_ccs.size(); i++){
        std::vector<long unsigned int> ccs = all_ccs[i];
        for (long unsigned int j=0; j<ccs.size(); j++){
            output += (std::to_string(ccs[j]) + "\t");
        }
        output+="\n";
    }
    std::ofstream f(OUTPUT_FILE);
    f<<output;
    f.close();

    return 0;
}